### PR TITLE
[Issue 97] Use higher-level WebSocketClient/Server classes, instead of *Frame

### PR DESCRIPTION
**Pull request by jhpedemonte**
Use `WebSocketServer` and `WebSocketClient` to do WS proxying, instead of relying on `http-proxy`.

Ref #97
Alternate to #180

### CODE REVIEW

**Comment by jhpedemonte** [comment]
This is a different take than #180. Instead of using `http-proxy` to do websocket proxying, this attempts to create a WS server (which connects to browser) and a WS client (which connects to kernel gateway) and proxy data between the two.  If this worked, it would be relatively easy to add in our message rewriting.

However, it's not working for me. I don't see any messages coming in (when running `make dev-logging`).

**Comment by jhpedemonte** [comment]
Sweet! Latest update makes things work. Tested on all of our example notebooks, although didn't run tests yet.  Still needs lots of cleanup.

@parente, please test to see if it fixes issue you were seeing with messages split across multiple WS frames.

**Comment by parente** [comment]
Will do.

**Comment by parente** [comment]
tl;dr: Works great.

Deployed the streaming meetup notebook to a dashboard server. Opened a browser tab to it. Started the stream. No problems for 5 minutes.

Decided to push my luck. Opened a second browser tab to the same dashboard. Started another stream. Walked away for 1.5 hours. My laptop may have suspended itself at some point. It at least locked itself.

Came back. Checked the consoles. No errors. One of them still streaming away happily. The other one apparently stopped. Toggle the stream on / off on that one got it going again immediately.

Websocket problem is gone.

**Comment by parente** [comment]
I know this PR isn't done, but one reminder: the kernel cleanup logic needs to be added back in. Probably got axed when the http proxy bit went away. It's probably significantly easier to implement now given that the websocket server probably emits a close/disconnect event when the browser goes away.

File: routes/api.js [hunk 1]
```diff
@@ -84,24 +88,103 @@ var substituteCodeCell = function(payload) {
...
+        });
+
+        client.on('connectFailed', function(e) {
+            error('ws client failure', e);
+        });
+        client.on('error', function(e) {
+            error('ws client error', e);
+        });
+        client.on('close', function() {
+            debug('closing WS client connection');
```
Comment (parente) [inline routes/api.js]: Probably need to move the kernel deletion timer and callback here. Doesn't seem to fire anymore on server close.

File: routes/api.js [hunk 2]
```diff
@@ -84,24 +88,103 @@ var substituteCodeCell = function(payload) {
...
+        // XXX verify request.origin
+        var servConn = request.accept(null, request.origin);
+
+        var pendingServMsgs = [];
+        var bufferServMsgs = function(data) {
+            pendingServMsgs.push(data);
+        };
+        servConn.on('message', bufferServMsgs);
+        servConn.on('close', function(reasonCode, desc) {
+            debug('closing WS server connection:', reasonCode, desc);
```
Comment (parente) [inline routes/api.js]: Probably need to move the kernel deletion timer and callback here. Doesn't seem to fire anymore on server close.

File: routes/api.js [hunk 3]
```diff
@@ -84,24 +88,103 @@ var substituteCodeCell = function(payload) {
...
+
+        var wsKgUrl = urljoin(kgUrl, kgBaseUrl, request.resourceURL.path).replace(/^http/, 'ws');
+        client.connect(wsKgUrl, null);
+    });
+
+    return;
+
+
     // Listen to the `upgrade` event and proxy the WebSocket requests as well.
     _server.on('upgrade', function(req, socket, head) {
```
Comment (parente) [inline routes/api.js]: I'm assuming this part goes away now? Are we sure it wasn't active somehow during the testing?

**Comment by jhpedemonte** [comment]
> I know this PR isn't done, but one reminder: the kernel cleanup logic needs to be added back in

Yep, Working on refactoring and making this pretty!

File: routes/api.js (hunk 3 above)
Comment (jhpedemonte) [inline routes/api.js]: Yes, goes away. The `upgrade` is now handled internally by the `WebSocketServer/Client` classes.

**Comment by jhpedemonte** [comment]
Latest commit passes integration tests. But unit tests have to be rewritten.

File: app/ws-rewriter.js [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+
+    wsserver.on('request', this._handleWsRequest.bind(this));
+}
+util.inherits(WsRewriter, EventEmitter);
+
+WsRewriter.prototype._handleWsRequest = function(req) {
+    debug('ws-server connection request');
+
+    var self = this;
+    var servConn = req.accept(null, req.origin); // accept from all origins
```
Comment (parente) [inline app/ws-rewriter.js]: Might need to think about XSRF if we're passing any origin through.

File: routes/api.js [hunk 5]
```diff
@@ -40,164 +36,50 @@ var proxy = httpProxy.createProxyServer({
...
-                        return ''; // throw away execute request
-                    });
-            }
-        } catch(e) {
-            // TODO better handle parse error in WS message
-            error('Failed to parse `data` in WS.', e);
-            transformedData = '';
-        }
+function initWsProxy(server) {
+    if (wsProxy) {
```
Comment (parente) [inline routes/api.js]: Are we still sharing one instance of the rewriter here?

**Comment by parente** [comment]
Tested with the docker-compose recipe by running 2 instances of the streaming meetup demo simultaneously. No websocket disconnects or other problems noted. Ran all the unit and integration tests and no problems there either.

Few minor questions inline above. Otherwise, good stuff. 🏆  Let's :shipit:

File: app/ws-rewriter.js (hunk 4 above)
Comment (jhpedemonte) [inline app/ws-rewriter.js]: I was wondering about that, but not sure what we would do here. This is the websocket server, so it has to accept incoming connections. It can't really know ahead of time who the client will be -- unless you have something else in mind?

File: routes/api.js (hunk 5 above)
Comment (jhpedemonte) [inline routes/api.js]: I think we should be fine. There is only one instance, but it sets up the websocket server such that on each `request`, it creates a new client to the backend and then proxies those connections. And each new connecting client generates a new `request`.

I think I tested this early at one point, but would need to do so again with latest code.

File: app/ws-rewriter.js (hunk 4 above)
Comment (parente) [inline app/ws-rewriter.js]: Nothing to solve here. We'd solve it at the auth level. Something akin to https://github.com/expressjs/csurf where we render a token onto the HTML page of the dashboard and expect to see that token back in the request.

We might already be getting something like XSRF protection from the way we're using session ID. Another security review is probably in order. For now, leave it as-is.

File: routes/api.js (hunk 5 above)
Comment (parente) [inline routes/api.js]: I tested with multiple clients and studied the code a bit more after I commented. I believe you're right that the only singleton is the WS server that accepts connections and spawns of instances of client/server to handle them independently. I couldn't see in the code any state that would be incorrectly shared across requests nor could I find a case in testing that causes crosstalk across clients. 👍
